src/utils.ts

```diff
@@ -1254,3 +1255,28 @@ export async function request(
     req.end();
   });
 }
+
+/**
+ * A custom AbortController that aborts after a specified timeout.
+ *
+ * If `timeout` is undefined or \<=0, the abort controller never aborts.
+ *
+ * This class provides two benefits over the built-in AbortSignal.timeout() method.
+ * - This class provides a mechanism for cancelling the timeout
+ * - This class supports infinite timeouts by interpreting a timeout of 0 as infinite.  This is
+ *    consistent with existing timeout options in the Node driver (serverSelectionTimeoutMS, for example).
+ * @internal
+ */
+export class TimeoutController extends AbortController {
+  constructor(
+    timeout?: number,
+    private timeoutId = timeout && timeout > 0 ? setTimeout(() => this.abort(), timeout) : null
+  ) {
```

Comment:
Should all the timers we make be .unref()-ed? is there a scenario where this timer could outlast the resources that are actually meaningful to keeping the eventloop turning (sockets)?

Reply:
No - when we close a client, we'll clear both the server selection queue and the connection checkout queue, cancelling all timeouts.